tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,744 @@
+//! to the task at the front of the queue, waking that task if its requested
+//! number of permits is met. 
+//!
+//! Because waiters are enqueued at the back of the linked list and dequeued
+//! from the front, the semaphore is fair. Tasks trying to acquire large numbers
+//! of permits at a time will always be woken eventually, even if many other
+//! tasks are acquiring smaller numbers of permits. This means that in a
+//! use-case like tokio's read-write lock, writers will not be starved by
+//! readers.
+//!
+//! The linked list is guarded by a mutex, which must be acquired before
+//! enqueueing or dequeueing a task. However, some operations are often
+//! wait-free.
+use crate::loom::cell::CausalCell;
+use crate::loom::sync::{atomic::AtomicUsize, Mutex, MutexGuard};
+use crate::util::linked_list::{self, LinkedList};
+
+use std::{cmp, fmt};
+use std::future::Future;
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+use std::ptr::NonNull;
+use std::sync::atomic::Ordering::*;
+use std::task::{Context, Poll, Waker};
+use std::task::Poll::*;
+
+/// An asynchronous counting semaphore which permits waiting on multiple permits at once.
+pub(crate) struct Semaphore {
+    waiters: Mutex<Waitlist>,
+    /// The current number of available permits in the semaphore.
+    permits: AtomicUsize,
+
+    /// Permits in the process of being released.
+    ///
+    /// When releasing permits, if the lock on the semaphore's wait list is held
+    /// by another task, the task releasing permits adds them to this counter.
+    /// The task holding the lock will continue releasing permits until the
+    /// counter is drained, allowing the `release` operation to be wait free.
+    ///
+    /// The first bit of this counter indicates that the semaphore is closing.
+    /// Therefore, all values are shifted over one bit.
+    adding: AtomicUsize,
+}
+
+struct Waitlist {
+    queue: LinkedList<Waiter>,
+    closed: bool,
+}
+
+/// Error returned by `Permit::try_acquire`.
+#[derive(Debug)]
+pub(crate) enum TryAcquireError {
+    Closed,
+    NoPermits,
+}
+/// Error returned by `Permit::poll_acquire`.
+#[derive(Debug)]
+pub(crate) struct AcquireError(());
+
+#[derive(Debug)]
+pub(crate) struct Permit {
+    state: PermitState,
+}
+
+pub(crate) struct Acquire<'a> {
+    node: Waiter,
+    semaphore: &'a Semaphore,
+    permit: &'a mut Permit,
+    num_permits: u16,
+}
+
+/// Permit state
+#[derive(Debug, Copy, Clone)]
+enum PermitState {
+    /// Currently waiting for permits to be made available and assigned to the
+    /// waiter.
+    Waiting(u16),
+
+    /// The number of acquired permits
+    Acquired(u16),
+}
+
+/// An entry in the wait queue.
+struct Waiter {
+    /// The current state of the waiter. 
+    /// 
+    /// This is either the number of remaining permits required by
+    /// the waiter, or a flag indicating that the waiter is not yet queued.
+    state: AtomicUsize,
```

Comment:
What are the benefits to having this be an atomic vs. a `usize` guarded by the mutex?

Reply:
Since my comment is still here, the benefit is that it avoids unsafe code.